src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/FindFeatureClustering.cpp

```diff
@@ -0,0 +1,297 @@
+  float32 r = 0.0f;
+
+  int32 bin = 0;
+  int32 ensemble = 0;
+  int32 totalPptFeatures = 0;
+  float32 min = std::numeric_limits<float32>::max();
+  float32 max = 0.0f;
+  float32 value = 0.0f;
+
+  std::vector<std::vector<float32>> clusters;
+  std::vector<float32> oldCount(m_InputValues->NumberOfBins);
+  std::vector<float32> randomRDF;
+
+  const usize totalFeatures = featurePhases.getNumberOfTuples();
+
+  SizeVec3 dims = imageGeometry.getDimensions();
+  FloatVec3 spacing = imageGeometry.getSpacing();
+
+  const float32 sizeX = dims[0] * spacing[0];
+  const float32 sizeY = dims[1] * spacing[1];
+  const float32 sizeZ = dims[2] * spacing[2];
+
+  // initialize boxDims and boxRes vectors
+  const std::array<float32, 3> boxDims = {sizeX, sizeY, sizeZ};
+  const std::array<float32, 3> boxRes = {spacing[0], spacing[1], spacing[2]};
+
+  for(usize i = 1; i < totalFeatures; i++)
+  {
+    if(featurePhases[i] == m_InputValues->PhaseNumber)
+    {
+      totalPptFeatures++;
+    }
+  }
+
+  clusters.resize(totalFeatures);
+
+  for(usize i = 1; i < totalFeatures; i++)
+  {
+    if(featurePhases[i] == m_InputValues->PhaseNumber)
+    {
+      if(i % 1000 == 0)
+      {
+        m_MessageHandler(IFilter::Message::Type::Info, fmt::format("Working on Feature {} of {}", i, totalPptFeatures));
+      }
+
+      x = centroids[3 * i];
+      y = centroids[3 * i + 1];
+      z = centroids[3 * i + 2];
+
+      for(usize j = i + 1; j < totalFeatures; j++)
+      {
+        if(featurePhases[i] == featurePhases[j])
+        {
+          xn = centroids[3 * j];
+          yn = centroids[3 * j + 1];
+          zn = centroids[3 * j + 2];
+
+          r = sqrtf((x - xn) * (x - xn) + (y - yn) * (y - yn) + (z - zn) * (z - zn));
+
+          clusters[i].push_back(r);
+          clusters[j].push_back(r);
+        }
+      }
+    }
+  }
+
+  for(usize i = 1; i < totalFeatures; i++)
+  {
+    for(usize j = 0; j < clusters[i].size(); j++)
+    {
+      if(featurePhases[i] == m_InputValues->PhaseNumber)
+      {
+        value = clusters[i][j];
+        if(value > max)
+        {
+          max = value;
+        }
+        if(value < min)
+        {
+          min = value;
+        }
+      }
+    }
+  }
+
+  const float32 stepSize = (max - min) / m_InputValues->NumberOfBins;
+
+  minMaxDistances[(m_InputValues->PhaseNumber * 2)] = max;
+  minMaxDistances[(m_InputValues->PhaseNumber * 2) + 1] = min;
+
+  for(usize i = 1; i < totalFeatures; i++)
+  {
+    for(usize j = 0; j < clusters[i].size(); j++)
+    {
+      if(featurePhases[i] == m_InputValues->PhaseNumber)
+      {
+        if(m_InputValues->RemoveBiasedFeatures && (*biasedFeatures)[i])
+        {
+          continue;
+        }
+
+        ensemble = featurePhases[i];
+        bin = (clusters[i][j] - min) / stepSize;
+        if(bin >= m_InputValues->NumberOfBins)
+        {
+          bin = m_InputValues->NumberOfBins - 1;
+        }
+        rdf[(m_InputValues->NumberOfBins * ensemble) + bin]++;
+      }
+    }
```

Comment:
```suggestion
    if(featurePhases[i] == m_InputValues->PhaseNumber)
    {
      if(m_InputValues->RemoveBiasedFeatures && (*biasedFeatures)[i])
      {
        continue;
      }
      for(usize j = 0; j < clusters[i].size(); j++)
      {
        ensemble = featurePhases[i];
        bin = (clusters[i][j] - min) / stepSize;
        if(bin >= m_InputValues->NumberOfBins)
        {
          bin = m_InputValues->NumberOfBins - 1;
        }
        rdf[(m_InputValues->NumberOfBins * ensemble) + bin]++;
      }
    }
```
Feature level checks outside the internal cluster for loop